Align __all__ and reexports with official documentation for linter compatibility

besynnerlig opened the pull request:
# Align **all** and reexports with official documentation for linter compatibility

This PR makes the documented import style (as shown in the HABApp docs)
fully compatible with type checkers such as Pylance, by aligning the
`__all__` definitions and using explicit `import X as X` reexports.

### Changes

-   **HABApp/**init**.py**: Added explicit reexports for `Parameter` and
    `DictParameter`.
-   **HABApp/core/events/**init**.py**: Explicit reexports and `__all__`
    for events and filters.
-   **HABApp/openhab/events/**init**.py**: Explicit reexports and
    `__all__` for OpenHAB events.
-   **HABApp/rule/scheduler/**init**.py**: Added `__all__` to expose
    `trigger`, `filter`, and related symbols.
-   **HABApp/openhab/items/**init**.py**: Explicit reexports and
    `__all__` for documented item classes.

### Motivation

-   The HABApp documentation encourages imports such as:

    ``` python
    from HABApp.core.events import ValueChangeEventFilter
    from HABApp.openhab.items import SwitchItem
    ```

-   These imports work at runtime, but Pylance raises
    `reportPrivateImportUsage` warnings without explicit reexports.

-   With these changes, the documented imports remain the recommended
    and future-proof style, while also producing clean linter output
    out-of-the-box.

### Notes

-   No functional behavior is changed.
-   Only reexports and `__all__` definitions are adjusted.
-   More involved type hint issues (e.g. ParamSpec in scheduler,
    `oh_set_value` aliasing) are left for a separate PR to keep this
    change minimal.

besynnerlig committed:
Make openhab.items reexports explicit and define __all__ for documented imports

besynnerlig committed:
Make openhab.events reexports explicit and add __all__ for linter support

besynnerlig committed:
Add explicit reexports for Parameter and DictParameter in HABApp root

besynnerlig committed:
Make core.events reexports explicit and add __all__ for Pylance compatibility

besynnerlig committed:
Add __all__ to scheduler module to expose trigger, filter and related symbols

spacemanspiff2007 commented on src/HABApp/__init__.py:
```diff
@@ -13,16 +13,21 @@
+    "Parameter",
+    "DictParameter",
+    "CONFIG",
+]
```
Is the reexport (``<name> as <name>``) enough to make pylance happy?
I'd prefer not to have to manually keep `__all__` in sync and I'd also like to discourage the usage of `from ... import *`

besynnerlig replied:
Thanks for the quick review and for shepherding this, @spacemanspiff2007, really appreciate your guidance and the careful API hygiene you keep in HABApp. 🙏

## What changed

- In the changed files, I previously switched to *explicit re-exports* using the `X as X` identity pattern, e.g.
  ```py
  from .base_item import BaseItem as BaseItem
  from .base_valueitem import BaseValueItem as BaseValueItem
  from .item import Item as Item
  from .item_aggregation import AggregationItem as AggregationItem
  from .item_color import ColorItem as ColorItem
  ```
  This is actually enough for Pylance/pyright to recognize these as public symbols without needing to maintain `__all__`. It also discourages `from … import *`, which I know you prefer to avoid. Runtime behavior is unchanged; this only improves editor/linter UX.

- **Removed `__all__`** everywhere **except one place**: `HABApp.rule.scheduler`.
  - In that module we re-export *third-party* symbols from `eascheduler` **and** we give them new public aliases (`trigger`, `filter`). In this specific aggregator+rename case, Pylance insists on an explicit public export list; if `__all__` is omitted, it flags:
    > "trigger" is not exported from module "HABApp.rule.scheduler" — import from "eascheduler.builder.triggers" instead
  - Keeping `__all__` *only there* makes the linter happy while preserving the nice public API `from HABApp.rule.scheduler import trigger`.
  - Everywhere else, explicit re-exports (`X as X`) were sufficient, so `__all__` was removed to avoid maintenance overhead and to discourage star-imports.

## What I didn’t include (to keep this PR focused on imports)

- There are still Pylance warnings around **`self.run.once/at`** and similar due to `ParamSpec` inference on bound methods:
  ```
  Arguments for ParamSpec "HINT_CB_P@at/once" are missing
  ```
  That’s not caused by imports, so I’ll handle it in a separate PR if you don't disagree... The proposed fix (minimal and back-compat) is to add two overloads alongside the current `ParamSpec` signature:
  - `Callable[[], Any]` (no-arg callback)
  - `Callable[[Any], Any]` (single-arg callback, e.g., passing `None`)
  and keep the existing generic `ParamSpec` definition as the fallback. This makes Pylance resolve the common cases without weakening the type safety you put in place.

If you’d like, I can include the overloads right away, but I kept this PR narrowly scoped so we can merge the import/export improvements cleanly first.

Happy to adjust anything to your taste, thanks again for the review and for keeping HABApp’s public surface crisp!

spacemanspiff2007 commented:
Thank you very much for your PR. This looks good, just a small remark and then it should be good to be merged.

besynnerlig committed:
Explicit re-exports for linter compatibility

besynnerlig committed:
Explicit re-exports for linter compatibility (drop __all__)

besynnerlig committed:
Explicit re-exports for linter compatibility (drop __all__)

besynnerlig committed:
Explicit re-exports for linter compatibility (drop __all__)

besynnerlig committed:
Explicit re-exports for linter compatibility (drop __all__)

besynnerlig committed:
style(rule.scheduler): normalize __all__ quotes to single (ruff compliance)

Use single quotes in HABApp.rule.scheduler.__init__.__all__ to match ruff’s formatting rules.
Keeps __all__ only where required for public re-exports of third-party symbols (trigger/filter).
No functional changes.

spacemanspiff2007 commented on src/HABApp/rule/scheduler/__init__.py:
```diff
@@ -1,6 +1,29 @@
-from HABApp.core.lib import InstantView
+from HABApp.core.lib import InstantView as InstantView
+
+__all__ = [
+    'trigger',
+    'filter',
+    'hours',
+    'minutes',
+    'seconds',
+    'milliseconds',
+    'add_holiday',
+    'get_holiday_name',
+    'get_holidays_by_name',
+    'get_sun_position',
+    'is_holiday',
+    'pop_holiday',
+    'InstantView',
+]
```
Thanks for making all checks pass. Could you remove this `__all__` block, too?

besynnerlig replied:
I can if you wish but I kept `__all__` only in `HABApp.rule.scheduler.__init__` because that module re-exports third-party symbols under new public names (`trigger`, `filter`). Without `__all__` there, Pylance flags:

> “`trigger` is not exported from module `HABApp.rule.scheduler` — import from `eascheduler.builder.triggers` instead.”

Keeping `__all__` in that one place preserves the existing

```py
from HABApp.rule.scheduler import trigger
```

UX and silences the warning.
Removing `__all__` in that specific file will reintroduce the linter errors.

spacemanspiff2007 replied:
Ah - I was not aware of that.
Okay, so let's keep `__all__` but do we then still need the reexport with `as` ?

besynnerlig replied:
Thanks for the quick follow-up!  

**Short answer:** with `__all__` present, the explicit `as` re-exports aren’t needed. I switched to plain imports and kept the `__all__` only in `HABApp.rule.scheduler.__init__` for the renamed third-party symbols (`trigger`, `filter`). This preserves `from HABApp.rule.scheduler import trigger` and keeps pyright/pylance happy.  

I also added a comment in the source to explain why `__all__` stays here.  

No functional changes — feel free to squash or adjust if you prefer a different approach.

besynnerlig committed:
style(rule.scheduler): keep __all__ and drop redundant re-export aliases

Use minimal surface in HABApp.rule.scheduler.__init__:
- keep __all__ for public API (trigger, filter, time helpers, holiday helpers, InstantView)
- remove redundant “as X” re-exports; plain imports + __all__ suffice for Pylance
No functional changes.

spacemanspiff2007 commented:
Congratulations on your first contribution!
